pkg/vm/engine/tae/containers/cn_vector_adapter.go

```diff
@@ -0,0 +1,645 @@
+			od := vec.Get(i).(types.Rowid)
+			if types.CompareRowidRowidAligned(d, od) != 0 {
+				return false
+			}
+		} else {
+			if vec.Get(i) != o.Get(i) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (vec *CnTaeVector[T]) ForeachWindow(offset, length int, op ItOp, sels *roaring.Bitmap) (err error) {
+	err = vec.forEachWindowWithBias(offset, length, op, sels, 0)
+	return
+}
+
+func (vec *CnTaeVector[T]) forEachWindowWithBias(offset, length int, op ItOp, sels *roaring.Bitmap, bias int) (err error) {
+	if sels == nil || sels.IsEmpty() {
+		for i := offset; i < offset+length; i++ {
+			elem := vec.Get(i + bias)
+			if err = op(elem, i); err != nil {
+				break
+			}
+		}
+	} else {
+
+		selsArray := sels.ToArray()
+		end := offset + length
+		for _, rowId := range selsArray {
+			if int(rowId) < offset {
+				continue
+			} else if int(rowId) >= end {
+				break
+			}
+			elem := vec.Get(int(rowId) + bias)
+			if err = op(elem, int(rowId)); err != nil {
+				break
+			}
+		}
+	}
+	return
+}
+
+func (vec *CnTaeVector[T]) PPString(num int) string {
+	var w bytes.Buffer
+	_, _ = w.WriteString(fmt.Sprintf("[T=%s][Len=%d][Data=(", vec.GetType().String(), vec.Length()))
+	limit := vec.Length()
+	if num > 0 && num < limit {
+		limit = num
+	}
+	size := vec.Length()
+	long := false
+	if size > limit {
+		long = true
+		size = limit
+	}
+	for i := 0; i < size; i++ {
+		if vec.IsNull(i) {
+			_, _ = w.WriteString("null")
+			continue
+		}
+		if vec.GetType().IsVarlen() {
+			_, _ = w.WriteString(fmt.Sprintf("%s, ", vec.Get(i).([]byte)))
+		} else {
+			_, _ = w.WriteString(fmt.Sprintf("%v, ", vec.Get(i)))
+		}
+	}
+	if long {
+		_, _ = w.WriteString("...")
+	}
+	_, _ = w.WriteString(")]")
+	return w.String()
+}
+
+//TODO: --- Pretty sure the below functions works. But need some validation
+
+func (vec *CnTaeVector[T]) Reset() {
+	if vec.Length() == 0 {
+		return
+	}
+
+	if vec.Nullable() {
+		cnNulls.Reset(vec.downstreamVector.Nsp)
+		//NOTE: We are not resetting the isNullable.
+	}
+
+	cnVector.Reset(vec.downstreamVector)
+}
+
+func (vec *CnTaeVector[T]) Close() {
+	vec.releaseDownstream()
+}
+
+func (vec *CnTaeVector[T]) releaseDownstream() {
+	if vec.downstreamVector != nil && vec.isAllocatedFromMpool {
+		vec.downstreamVector.Free(vec.mpool)
+		vec.downstreamVector = nil
+		vec.isAllocatedFromMpool = false
+	}
+}
+
+func (vec *CnTaeVector[T]) Allocated() int {
+
+	//1. If mpool doesn't allocate
+	if !vec.isAllocatedFromMpool {
+		return 0
+	}
+
+	// 2. if mpool allocates
+	{
+		if vec.GetType().IsVarlen() {
+			return vec.downstreamVector.Size()
+		}
+
+		return vec.downstreamVector.Length()
+	}
+}
+
+// TODO: --- I am not sure, if the below functions will work as expected
+
+func (vec *CnTaeVector[T]) Capacity() int {
+	// TODO: Can we use Length() instead of Capacity?
+	// TODO: Not used much. Can we remove?
+
+	// TODO: Capacity should be based on a number and not based on the Length. Fix it later.
+	return vec.Length()
```

Comment:
It is the maximum capacity of a vector. The difference with the Allocated:
1. type sensitive
2. For a read-only vector, Allocted always returns 0 but Capacity doesn't.

Reply:
Ok. I have some questions

1. Is it correct to use `vec.Length()` here? 
2. Or should it be `cap(vec.downstream.data)`? (Failed some test cases last time)
3. Is `Capacity()` required? 
- I see that `vector.Capacity()` is only used in `Batch.Capacity()`, and `Batch.Capacity()` is not used anywhere else. 
- `stl.Vector.Capacity()` is only used in `buffer.go`, which is not used anywhere else.

Reply:
You can remove Capacity.

Reply:
Got it. Thank you.